src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -117,83 +95,103 @@ private static void writeFast_int(MemorySegment segment, long offset, Charset ch
         segment.set(JAVA_INT, offset + bytes.length, 0);
     }
 
-    private static int native_strlen_byte(MemorySegment segment, long start) {
-        // Heap segments must be handled by Java code
-        if (!segment.isNative()) {
-            return strlen_byte(segment, start);
-        }
-
-        if (start > 0) {
-            segment = segment.asSlice(start);
+    /**
+     * {@return the shortest distance beginning at the provided {@code start}
+     *  to the encountering of a zero byte in the provided {@code segment}}
+     * <p>
+     * The method divides the region of interest into three distinct regions:
+     * <ul>
+     *     <li>head (un-aligned access handling on a byte-by-byte basis) (if any)</li>
+     *     <li>body (long aligned access handling eight bytes at a time) (if any)</li>
+     *     <li>tail (un-aligned access handling on a byte-by-byte basis) (if any)</li>
+     * </ul>
+     * <p>
+     * The body is using a heuristic method to determine if a long word
+     * contains a zero byte. The method might have false positives but
+     * never false negatives.
+     * <p>
+     * This method is inspired by the `glibc/string/strlen.c` implementation
+     *
+     * @param segment to examine
+     * @param start   from where examination shall begin
+     * @throws IllegalArgumentException if the examined region contains no zero bytes
+     *                                  within a length that can be accepted by a String
+     */
+    public static int chunked_strlen_byte(MemorySegment segment, long start) {
+
+        // Handle the first unaligned "head" bytes separately
+        int headCount = (int)SharedUtils.remainsToAlignment(segment.address() + start, Long.BYTES);
```

Comment:
This logic still seems to be tied to native segments. Heap segments have virtualized addresses, so this logic seems invalid. The best we can do, for heap segments, is to only use the better algo if we're using a long[] as an heap base (in which case the logic here would work reliably, and turn into aligned reads).